tests/std/tests/GH_001858_iostream_exception/test.cpp

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <cstring>
+#include <ios>
+#include <istream>
+#include <ostream>
+#include <streambuf>
+
+using namespace std;
+
+template <class CharT>
+class throwing_buffer : public basic_streambuf<CharT> {
+public:
+    streampos seekoff(streamoff, ios_base::seekdir, ios_base::openmode = ios_base::in | ios_base::out) override {
+        throw 42;
+    }
+
+    streampos seekpos(streampos, ios_base::openmode = ios_base::in | ios_base::out) override {
+        throw 42;
+    }
+
+    int sync() override {
+        throw 42;
+    }
+
+    basic_streambuf<CharT>* to_buf() {
+        return this;
+    }
+};
+
+template <class CharT>
+void test_istream_exceptions() {
+    throwing_buffer<CharT> buffer;
+
+    { // sync
+        basic_istream<CharT> is(buffer.to_buf());
+        assert(!is.bad());
+        assert(is.sync() == -1);
+        assert(is.bad());
+    }
+
+    { // sync with exceptions
+        basic_istream<CharT> is(buffer.to_buf());
+        is.exceptions(ios_base::badbit);
+
+        try {
+            assert(is.sync() == -1);
+        } catch (const ios_base::failure&) {
+            assert(false);
+        } catch (...) {

```

Comment:
Suggestion: `catch (...)` doesn't indicate that we're expecting to catch `throw 42;`. `catch (const int&)` would be more specific, although I believe that a `test_exception` would be the most clear.

Occurs throughout.